### PR TITLE
Stats per track

### PR DESCRIPTION
**Pull request by ashellunts**
New tests for telemetry + implementation of per track statistics.

### CODE REVIEW

**Review by boks1971 — COMMENTED** [review]
Looks great @ashellunts !!! Nice work!

Left a few comments. But, I am not hitting approve as I am not fully familiar with this part of the code and more 👀 would be nice.

Also, the `buffer` is per layer of simulcast. These stats are going to aggregate across simulcast layer if I understand right. I think it is worth doing per layer per track stats. Not sure if you all have already discussed that level of granularity and decided against it or do it later.

**Comment by ashellunts** [comment]
> Looks great @ashellunts !!! Nice work!
> 
> Left a few comments. But, I am not hitting approve as I am not fully familiar with this part of the code and more 👀 would be nice.
> 
> Also, the `buffer` is per layer of simulcast. These stats are going to aggregate across simulcast layer if I understand right. I think it is worth doing per layer per track stats. Not sure if you all have already discussed that level of granularity and decided against it or do it later.

Thank you for review! I did not realise buffers are different per layer. Current implementation only tracks bytes for last reported buffer. I would need to change implementation to handle them correctly.

For upstream, I looked how to differentiate RTCP stats, but looks like there is only 1 mediaTrack through which RTCP is reported and track.ssrc is ssrc of minimal layer.

In general I would prefer to finish per stream stats and merge it so that we have working version. After we can discuss if we want per layer stats.

File: pkg/telemetry/test/telemetry_service_test.go [hunk 1]
```diff
@@ -23,7 +25,32 @@ func createFixture() *telemetryServiceFixture {
...
+	//prepare
+	room := &livekit.Room{Sid: "RoomSid", Name: "RoomName"}
+	partSID := "part1"
+	clientInfo := &livekit.ClientInfo{Sdk: 2}
+	participantInfo := &livekit.ParticipantInfo{Sid: partSID}
+	fixture.sut.ParticipantJoined(context.Background(), room, participantInfo, clientInfo)
+
+	//do
+	packet := 33
+	fixture.sut.OnDownstreamPacket(partSID, "", packet)
```
[the remark this answers is not in the document]
Comment (ashellunts) [reply pkg/telemetry/test/telemetry_service_test.go:40]: Right, this test is only to check that roomID/name and participant ID are sent in analytics. Track ID stuff is handled in different tests.

File: pkg/telemetry/test/telemetry_service_test.go [hunk 2]
```diff
@@ -49,6 +77,248 @@ func Test_OnDownstreamPacket(t *testing.T) {
...
+		&rtcp.ReceiverReport{
+			Reports: []rtcp.ReceptionReport{
+				{Delay: 1, Jitter: 5, TotalLost: 3},
+				{Delay: 4, Jitter: 2, TotalLost: 4},
+			},
+		},
+	}
+	// there should be bytes reported so that stats are sent
+	buf := &buffer.Buffer{}
+	buf.SetStats(buffer.Stats{
```
[the remark this answers is not in the document]
Comment (ashellunts) [reply pkg/telemetry/test/telemetry_service_test.go:232]: ok, will rename to `SetStatsTestOnly`

File: pkg/telemetry/statsworker.go [hunk 3]
```diff
@@ -43,48 +42,67 @@ func newStatsWorker(ctx context.Context, t TelemetryReporter, roomID, roomName,
...
+func (s *StatsWorker) createIncomingStatsIfEmpty(trackID string) *Stats {
+	if s.incomingPerTrack[trackID] == nil {
+		s.incomingPerTrack[trackID] = &Stats{next: &livekit.AnalyticsStat{
+			Kind:          livekit.StreamType_UPSTREAM,
+			RoomId:        s.roomID,
+			ParticipantId: s.participantID,
+			RoomName:      s.roomName,
+		}}
+	}
+	return s.outgoingPerTrack[trackID]
```
[the remark this answers is not in the document]
Comment (ashellunts) [reply pkg/telemetry/statsworker.go:91]: fixed

File: pkg/telemetry/statsworker.go [hunk 4]
```diff
@@ -43,48 +42,67 @@ func newStatsWorker(ctx context.Context, t TelemetryReporter, roomID, roomName,
...
+func (s *StatsWorker) createOutgoingStatsIfEmpty(trackID string) *Stats {
+	if s.outgoingPerTrack[trackID] == nil {
+		s.outgoingPerTrack[trackID] = &Stats{next: &livekit.AnalyticsStat{
+			Kind:          livekit.StreamType_DOWNSTREAM,
+			RoomId:        s.roomID,
+			ParticipantId: s.participantID,
+			RoomName:      s.roomName,
+		}}
 	}
+	return s.outgoingPerTrack[trackID]
```
[the remark this answers is not in the document]
Comment (ashellunts) [reply pkg/telemetry/statsworker.go:78]: I wanted to use first, and decided not to use, that is why it is confusing now.

Fixed by using `getOrCreate` pattern.

File: pkg/telemetry/statsworker.go [hunk 5]
```diff
@@ -43,48 +42,67 @@ func newStatsWorker(ctx context.Context, t TelemetryReporter, roomID, roomName,
...
+}
+
+func (s *StatsWorker) OnRTCP(trackID string, direction livekit.StreamType, stats *livekit.AnalyticsStat) {
+	s.Lock()
+	defer s.Unlock()
+
+	var ds *Stats
+	if direction == livekit.StreamType_DOWNSTREAM {
+		s.createOutgoingStatsIfEmpty(trackID)
```
[the remark this answers is not in the document]
Comment (ashellunts) [reply pkg/telemetry/statsworker.go]: done

File: pkg/rtc/mediatrack.go [hunk 6]
```diff
@@ -631,6 +631,9 @@ func (t *MediaTrack) handlePublisherFeedback(packets []rtcp.Packet) {
 	var totalLost uint32
 	var maxSeqNum uint32
 
+	//forward to telemetry
+	t.params.Telemetry.HandleRTCP(livekit.StreamType_UPSTREAM, t.params.ParticipantID, t.ID(), packets)
```
Comment (frostbyte73) [inline pkg/rtc/mediatrack.go:635]: Does this include padding packets? Not sure if it changed, but this was giving me a lower packet count and byte count than the interceptor

Comment (ashellunts) [reply pkg/rtc/mediatrack.go:635]: This is not used for packet count and byte count. They are received from buffer.Buffer as before. RTCP values are same as from interceptor, I have checked that.

**Review by davidzhao — APPROVED** [review]

**Comment by ashellunts** [comment]
Merging now, @davidzhao thank you for approve. @frostbyte73 please let me know if you have any other feedback for the PR.

**Comment by frostbyte73** [comment]
> Merging now, @davidzhao thank you for approve. @frostbyte73 please let me know if you have any other feedback for the PR.

Nope, good to go

**Pull request merged**